# PSMC

## Creating artifitial diplod individual

We concatnated the fastq file of DNAseq from two individual into two fastq file R1.fastq and R2.fastq.

Then the concatened paired reads R1 and R2 are aligned to species reference genome in the same way as shown in dN/dS DNAseq alignment [workflow.py](https://github.com/Jilong-Jerome/sociality-in-spiders-dead-end/blob/main/dNdS/dna_align/workflow.py)

## Prepare PSMC input
1. Generate consensus fastq for each chromsome separately

2. Convert consensus fastq to psmcfa format for psmc

3. Combine psmcfa from chromosmes into autosomes set and X chromosomes set

## Run PSMC

## Filling values and calculating Ne_X/Ne_A



```
something
```
